lib/decompress/zstd_decompress.c

```diff
@@ -544,6 +556,59 @@ static size_t readSkippableFrameSize(void const* src, size_t srcSize)
     }
 }
 
+/*! ZSTD_findFrameSize() :
+ * Returns the raw size of the first frame in a buffer.  Advancing the pointer by this amount will move to the next frame.
+ *  `src` should point to the start of a buffer with one or more ZSTD encoded frames.
+ *  `srcSize` must be at least as large as the frame header.
+ *            hint : any size >= `ZSTD_frameHeaderSize_max` is large enough.
+ *  @return : - size of `src` of the first frame, if known.  Advancing the pointer by this amount will move to the next frame.
+ *            - ZSTD_CONTENTSIZE_UNKNOWN if the size cannot be determined
+ *            - ZSTD_CONTENTSIZE_ERROR if an error occurred (e.g. invalid magic number, srcSize too small)
+ */
+unsigned long long ZSTD_findFrameSize(const void *src, size_t srcSize)
+{
+    ZSTD_frameHeader zfh;
+    if (ZSTD_getFrameHeader(&zfh, src, srcSize) != 0)
+        return ZSTD_CONTENTSIZE_ERROR;
+    if (zfh.frameType == ZSTD_skippableFrame) {
+        return readSkippableFrameSize(src, srcSize);
+    } else {
+        return ZSTD_findFrameCompressedSize(src, srcSize);
+    }
+}
+
+/*! ZSTD_readSkippableFrame() :
+ * Retrieves a zstd skippable frame containing data given by src, and writes it to dst buffer.
+ *
+ * The parameter magicVariant will receive the magicVariant that was supplied when the frame was written,
+ * i.e. magicNumber - ZSTD_MAGIC_SKIPPABLE_START
+ *
+ * Returns an error if destination buffer is not large enough, if the source size is not representable
+ * with a 4-byte unsigned int, or if the frame is not skippable.
+ *
+ * @return : number of bytes written or a ZSTD error.
+ */
+
+ZSTDLIB_API unsigned long long ZSTD_readSkippableFrame(void* dst, size_t dstCapacity, unsigned* magicVariant,
+                                            const void* src, size_t srcSize)
+{
+    U32 const magicNumber = MEM_readLE32(src);
+    size_t skippableFrameSize = readSkippableFrameSize(src, srcSize);
+    size_t skippableContentSize = skippableFrameSize - ZSTD_SKIPPABLEHEADERSIZE;
+
+    /* check input validity */
+    RETURN_ERROR_IF(!ZSTD_isSkippableFrame(src, srcSize), frameParameter_unsupported, "");
+    if (skippableFrameSize < ZSTD_SKIPPABLEHEADERSIZE || skippableFrameSize > srcSize)
+        return ZSTD_CONTENTSIZE_ERROR;
+    if (skippableContentSize > dstCapacity)
+        return ZSTD_CONTENTSIZE_ERROR;
+
+    /* deliver payload */
+    ZSTD_memcpy(dst, (const BYTE *)src + ZSTD_SKIPPABLEHEADERSIZE, skippableContentSize);
```

Comment:
This will need to handle `dst == NULL` and `skippableContentSize == 0`, since it is UB to call memcpy() with `NULL.